contrib/nexusproto/serializer.go

```diff
@@ -0,0 +1,174 @@
+package nexusproto
+
+import (
+	"fmt"
+	"mime"
+	"reflect"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+func messageFromAny(v any) (proto.Message, error) {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("invalid protobuf object: type: %T", v)
+	}
+	elem := rv.Type().Elem()
+
+	// v is a proto.Message or *proto.Message, the serializer supports either:
+	// 1. var d durationpb.Duration; s.Deserialize(&d)
+	// 2. var d *durationpb.Duration; s.Deserialize(&d)
+	// In the second case, we need to instantiate an empty proto struct:
+	if elem.Kind() == reflect.Ptr {
+		empty := reflect.New(elem.Elem())
+		rv.Elem().Set(empty)
+		rv = empty
+	}
+
+	msg, ok := rv.Interface().(proto.Message)
+	if !ok {
+		return nil, fmt.Errorf("%w: value is not a proto.Message or a pointer to one", nexus.ErrSerializerIncompatible)
+	}
+
+	return msg, nil
+}
+
+type protoJsonSerializer struct{}
+
+func (protoJsonSerializer) extractMessageType(contentType string) string {
+	if contentType == "" {
+		return ""
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil || mediaType != "application/json" || len(params) != 2 || params["format"] != "protobuf" {
+		return ""
+	}
+	return params["message-type"]
+}
+
+func (s protoJsonSerializer) Deserialize(c *nexus.Content, v any) error {
+	messageTypeFromHeader := s.extractMessageType(c.Header["type"])
+	if messageTypeFromHeader == "" {
+		return fmt.Errorf("%w: incompatible content type header", nexus.ErrSerializerIncompatible)
+	}
```

Comment:
I thought the incompatible-serializer made a bit of sense here. Can you help me understand fall-through behavior on deserialize? It seems we don't pick the deserializer _explicitly_ based on the mime type, so we cycle through right? So if I have both json and binary enabled, but json preferred, does this error cause it to fall through to binary? Ideally we only want some errors to fall through (e.g. incompatible), but other errors to be bubbled immediately without trying other deserializers.

Reply:
Fallthrough is only on ErrSerializerIncompatible.

Reply:
Yup, that's what is happening here.

Reply:
Ah, I see, "Fallthrough is only on ErrSerializerIncompatible and any error with that in the chain"